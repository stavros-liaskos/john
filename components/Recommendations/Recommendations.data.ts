export const recommendationsI18n = {
  title: 'Other artists',
  artistList: {
    btnTxt: 'Unfollow',
    noArtists: 'You don not track any artists yet',
  },
};
