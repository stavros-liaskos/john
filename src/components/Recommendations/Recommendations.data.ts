export const recommendationsI18n = {
  title: 'Others follow',
  artistList: {
    btnTxt: 'Follow',
    noArtists: 'There are no recommended artists for you :(',
  },
};
