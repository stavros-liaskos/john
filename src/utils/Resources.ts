import { components } from '../types/schema';

interface IResources {
  cacheMap: Record<string, unknown>;
  fetch: (url: string) => void;
}

class Resources implements IResources {
  private cacheMap;

  constructor() {
    this.cacheMap = new Map();
  }

  fetch(url) {
    if (this.cacheMap.has(url)) {
      return this.cacheMap.get(url);
    }

    const promise = fetch(url, {
      method: 'GET',
      credentials: 'include',
    })
      .then(response => {
        return response.json();
      })
      .then((data: components['schemas']['FollowedArtistsResponse']) => {
        this.cacheMap.set(url, data);
        return data;
      });

    throw promise;
  }
}

const resources = new Resources();

export default resources;
