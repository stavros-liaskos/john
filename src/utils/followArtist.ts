import { components } from '../types/schema';
import Endpoints from '../types/endpoints';

async function followArtist(artist: components['schemas']['SearchResultArtistDto'], cb: () => void) {
  const headers = new Headers({ 'Content-Type': 'application/json' });
  await fetch(Endpoints.Follow, {
    method: 'POST',
    credentials: 'include',
    headers,
    body: JSON.stringify(artist),
  })
    .then(() => {
      console.log(`${artist.name} followed successfully`);
    })
    .catch(error => {
      console.error('Error:', JSON.stringify(error));
    })
    .finally(() => {
      cb();
    });
}
export default followArtist;
