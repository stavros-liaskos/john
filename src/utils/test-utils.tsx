import React from 'react';
import { render, act } from '@testing-library/react';
import ThemeProvider from '../contexts/Theme/ThemeProvider';
import { UserProvider } from '@auth0/nextjs-auth0/client';
import ArtistsListProvider from '../contexts/ArtistsList/ArtistsListProvider';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider>
      <UserProvider>
        <ArtistsListProvider>{children}</ArtistsListProvider>
      </UserProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: { [key: string]: unknown }) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const renderWithAct = async (ui: React.ReactElement, options?: { [key: string]: unknown }) => {
  const component = render(ui, { wrapper: AllTheProviders, ...options });

  const promise = Promise.resolve();
  await act(() => promise);

  return component;
};

export const beforeEachTest = () => {
  jest.resetModules();
  jest.clearAllMocks();
  jest.restoreAllMocks();
  localStorage.clear();
};
