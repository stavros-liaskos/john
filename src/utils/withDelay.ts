export default function withDelay(cb: () => void, delay = 1000) {
  return setTimeout(() => {
    cb();
  }, delay);
}
