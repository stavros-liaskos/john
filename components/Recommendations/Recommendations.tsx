import ArtistsList, { ArtistsListI18n } from '../ArtistsList/ArtistsList';
import React, { useEffect, useState } from 'react';
import { components } from '../../types/schema';
import { followArtist } from '../../utils/followArtist';

type RecommendationsI18n = {
  title: string;
  artistList: ArtistsListI18n;
};

const Recommendations = ({ i18n }: { i18n: RecommendationsI18n }) => {
  const [artistLoading, setArtistLoading] = useState<number>(0);
  const [recommendedArtists, setRecommendedArtists] = useState<components['schemas']['FollowedArtistDto'][]>([]);

  useEffect(() => {
    // TODO add Suspense for data fetching
    fetchRecommendations(setRecommendedArtists);
  }, []);

  if (!i18n || !i18n.title) {
    return null;
  }

  return (
    <div className="flex flex-col flex-2 basis-24 lg:justify-center items-center mb-2 w-full">
      <h3 className={'h3'}>{i18n.title}</h3>
      <ArtistsList
        i18n={i18n.artistList}
        artistsList={recommendedArtists}
        onButtonClick={handleFollow}
        artistLoading={artistLoading}
      />
    </div>
  );

  async function handleFollow(artist: components['schemas']['SearchResultArtistDto']) {
    artist?.id && setArtistLoading(artist.id);

    const finallyCb = () => {
      artistLoading && setArtistLoading(0);
    };
    await followArtist(artist, finallyCb);
  }
};

export function fetchRecommendations(
  setRecommendedArtists: React.Dispatch<React.SetStateAction<components['schemas']['FollowedArtistDto'][]>>,
) {
  fetch(`${process.env.BE_BASE_URL}/artists/recommended?page=0&size=10`, {
    method: 'GET',
    credentials: 'include',
  })
    .then(res => res.json())
    .then((result: components['schemas']['FollowedArtistsResponse']) => {
      result?.rows && setRecommendedArtists(result?.rows);
    })
    .catch(console.error);
}

export default Recommendations;
