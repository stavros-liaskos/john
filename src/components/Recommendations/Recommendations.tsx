import ArtistsList, { ArtistsListI18n } from '../ArtistsList/ArtistsList';
import React, { useState } from 'react';
import { components } from '../../types/schema';
import { followArtist } from '../../utils/followArtist';
import resources from '../../utils/Resources';

type RecommendationsI18n = {
  title: string;
  artistList: ArtistsListI18n;
};

const Recommendations = ({ i18n }: { i18n: RecommendationsI18n }) => {
  const [artistLoading, setArtistLoading] = useState<number>(0);
  const url = `${process.env.BE_BASE_URL}/artists/recommended?page=0&size=10`;

  const recommendedArtists = resources.fetch(url);

  if (!i18n || !i18n.title) {
    return null;
  }

  return (
    <div className="flex flex-col flex-2 basis-24 lg:justify-center items-center mb-2 w-full">
      <h3 className={'h3'}>{i18n.title}</h3>
      <ArtistsList
        i18n={i18n.artistList}
        artistsList={recommendedArtists.rows}
        onButtonClick={handleFollow}
        artistLoading={artistLoading}
      />
    </div>
  );

  async function handleFollow(artist: components['schemas']['SearchResultArtistDto']) {
    artist?.id && setArtistLoading(artist.id);

    const finallyCb = () => {
      artistLoading && setArtistLoading(0);
    };
    await followArtist(artist, finallyCb);
  }
};

Recommendations.whyDidYouRender = true;
export default Recommendations;
