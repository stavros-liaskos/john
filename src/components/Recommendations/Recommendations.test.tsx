import { act, render, waitFor } from '@testing-library/react';
import Recommendations from './Recommendations';
import { recommendationsI18n } from './Recommendations.data';
import recommendedArtists from '../../../mocks/responses/followed-artists.json';
import { beforeEachTest } from '../../utils/test-utils';
import followedArtists from '../../../mocks/responses/followed-artists.json';

describe('Recommendations', () => {
  let originFetch: {
    (input: RequestInfo | URL, init?: RequestInit | undefined): Promise<Response>;
  };
  beforeEach(() => {
    originFetch = global.fetch;

    beforeEachTest();
  });

  afterEach(() => {
    global.fetch = originFetch;
    jest.restoreAllMocks();
  });

  it('renders without data without crashing', () => {
    // @ts-ignore
    render(<Recommendations />);
  });

  it('renders title and artists', async () => {
    const mRes = { json: jest.fn().mockResolvedValueOnce(recommendedArtists) };
    const mockedFetch = jest.fn().mockResolvedValueOnce(mRes);
    global.fetch = mockedFetch;

    const { getByText, findAllByText } = render(<Recommendations i18n={recommendationsI18n} />);

    expect(getByText(recommendationsI18n.title)).toBeInTheDocument();

    expect(mockedFetch).toHaveBeenCalledTimes(1);
    await waitFor(() => {
      expect(mRes.json).toHaveBeenCalledTimes(1);
    });
    const buttons = await findAllByText(recommendationsI18n.artistList.btnTxt);
    expect(buttons).toHaveLength(2);
  });

  it('matches snapshot', async () => {
    const fakeResponse = followedArtists;
    const mRes = { json: jest.fn().mockResolvedValueOnce(fakeResponse) };
    const mockedFetch = jest.fn().mockResolvedValueOnce(mRes);
    global.fetch = mockedFetch;

    const { container } = render(<Recommendations i18n={recommendationsI18n} />);
    await act(() => {});

    expect(container).toMatchSnapshot();
  });
});
