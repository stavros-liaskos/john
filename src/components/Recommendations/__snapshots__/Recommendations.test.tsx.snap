// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommendations matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col lg:justify-center items-center mb-2 w-full"
  >
    <h3
      class="h3"
    >
      Others follow
    </h3>
    <div
      class="overflow-auto w-full flex-auto h-64"
    >
      <p
        class="rr-text flex justify-center"
      >
        There are no recommended artists for you :(
      </p>
    </div>
  </div>
</div>
`;
